Source/Particles/Collision/BinaryCollision/BinaryCollision.H

```diff
@@ -447,25 +470,98 @@ public:
+            amrex::ParticleReal* AMREX_RESTRICT KE_in_each_cell = KE_vec.dataPtr();
+            amrex::ParticleReal* AMREX_RESTRICT px_in_each_cell = px_vec.dataPtr();
+            amrex::ParticleReal* AMREX_RESTRICT py_in_each_cell = py_vec.dataPtr();
+            amrex::ParticleReal* AMREX_RESTRICT pz_in_each_cell = pz_vec.dataPtr();
+
+            // Save the momentum before the collision since sometimes the correction fails
+            // and the only solution is to restore the pre-collision values.
+            // The implicit solver already has ux_n etc so use those if available,
+            // otherwise create temporaries.
+            amrex::Gpu::DeviceVector<amrex::ParticleReal> u1x_before;
+            amrex::Gpu::DeviceVector<amrex::ParticleReal> u1y_before;
+            amrex::Gpu::DeviceVector<amrex::ParticleReal> u1z_before;
+            amrex::ParticleReal* AMREX_RESTRICT u1x_before_ptr = nullptr;
+            amrex::ParticleReal* AMREX_RESTRICT u1y_before_ptr = nullptr;
+            amrex::ParticleReal* AMREX_RESTRICT u1z_before_ptr = nullptr;
+            if (m_correct_energy_momentum) {
+                // Check if ux_n was added.
+                std::vector<std::string> const & real_names1 = species_1.GetRealSoANames();
+                auto const pos1 = std::find(real_names1.begin(), real_names1.end(), "ux_n");
+                if (pos1 != real_names1.end()) {
+                    int const u1x_ni = std::distance(real_names1.begin(), pos1);
+                    u1x_before_ptr = soa_1.m_runtime_rdata[u1x_ni];
+                    u1y_before_ptr = soa_1.m_runtime_rdata[u1x_ni + 1];
+                    u1z_before_ptr = soa_1.m_runtime_rdata[u1x_ni + 2];
+                } else {
+                    u1x_before.resize(np1);
+                    u1y_before.resize(np1);
+                    u1z_before.resize(np1);
+                    u1x_before_ptr = u1x_before.dataPtr();
+                    u1y_before_ptr = u1y_before.dataPtr();
+                    u1z_before_ptr = u1z_before.dataPtr();
+                }
+
+            }
+
+            amrex::ParticleReal const beta_weight_exponent = m_beta_weight_exponent;
+
             WARPX_PROFILE_VAR("BinaryCollision::doCollisionsWithinTile::findDensityTemperatures", prof_findDensityTemperatures);
-        WARPX_PROFILE_VAR("BinaryCollision::doCollisionsWithinTile::findDensityTemperatures::atomics", prof_findDensityTemperatures_atomics);
+
+            if (m_correct_energy_momentum) {
+                WARPX_PROFILE_VAR("BinaryCollision::doCollisionsWithinTile::findMomentumEnergy", prof_findMomentumEnergy);
+                amrex::ParallelFor( np1,
+                    [=] AMREX_GPU_DEVICE (int i1) noexcept
+                    {
+                        u1x_before_ptr[i1] = u1x[i1];
+                        u1y_before_ptr[i1] = u1y[i1];
+                        u1z_before_ptr[i1] = u1z[i1];
```

Comment:
Just for my education: if `ux/y/z_n` exist as particle attributes, this will overwrite those values. Is it the case that the values in those arrays are already equal to `u1x/y/z`?

Reply:
The `ux/y/z_n` are there for the implicit solvers which need to save the velocities at the start of the time steps. The arrays are only used during the implicit advance. Using them here won't interfere with the solver since the collisions are done separately from the solver. As it is now, the data is copied into the `ux/y/z_n` at the start of the solver, so this copy is needed here also to get the values after the advance.